test/app-luatest/key_estimator_test.lua

```diff
@@ -0,0 +1,187 @@
+local t = require('luatest')
+local g = t.group()
+local key_def = require('key_def')
+local key_estimator = require('key_estimator')
+
+local function hll_n_counters(prec)
+    return 2 ^ prec
+end
+
+--[[
+Get the standard error of the estimation of the HyperLogLog algorithm.
+The standard error determines the probability of deviation of the estimation
+from the actual value. ** let sigma = srd_err * actual_value **
+The deviation is less than 1*sigma with probability of 68.2%,
+                           2*sigma with probability of 95.4%,
+                           3*sigma with probability of 99.7%.
+]]--
+local function hll_std_error(prec)
+    local n_counters = hll_n_counters(prec)
+    return 1.04 / (n_counters ^ 0.5)
+end
+
+g.test_constants = function()
+    t.assert_not_equals(key_estimator.SPARSE, key_estimator.DENSE)
+    t.assert_lt(key_estimator.MIN_PRECISION, key_estimator.MAX_PRECISION)
+end
+
+g.test_common = function()
+    t.assert_error_msg_equals(
+        "Usage: key_estimator.new(format = <key_def>" ..
+        "[, precision = <num>), " ..
+        "representation = key_estimator.(SPARSE|DENSE)]",
+        key_estimator.new)
+
+    local fmt = key_def.new({{type = "string", fieldno = 1}})
+    local est = key_estimator.new(fmt)
+
+    t.assert_error_msg_equals(
+        "Invalid precision 999 " ..
+        "(available values are" ..
+        " from " .. key_estimator.MIN_PRECISION ..
+        " to " .. key_estimator.MAX_PRECISION .. ")",
+        key_estimator.new, fmt, 999)
+
+    t.assert_equals(est:estimate(), 0)
+    est:add({'bla'})
+    t.assert_equals(est:estimate(), 1)
+    est:add({'bla'})
+    t.assert_equals(est:estimate(), 1)
+    est:add({'bla', 'bla', 'bla..'})
+    t.assert_equals(est:estimate(), 1)
+    est:add({'another bla'})
+    t.assert_equals(est:estimate(), 2)
+
+    t.assert_error_msg_equals("Invalid tuple format.",
+        key_estimator.add, est, {0})
+end
+
+g.test_hll_dense_estimation_error = function()
+    local fmt = key_def.new({{type = "string", fieldno = 1}})
+
+    local min_prec = key_estimator.MIN_PRECISION
+    local max_prec = key_estimator.MAX_PRECISION
+    for prec = min_prec, max_prec, 1 do
+        local est = key_estimator.new(fmt, prec)
+        local std_err = hll_std_error(prec)
+
+        --- cardinality for wich the HyperLogLog algorithm will be used.
```

Comment:
No, by explaining the formula I mean, where does `5` come from?

Reply:
Added a more detailed description.